docs: Describe Otel log gateway PoC (#843)
kyma-project/telemetry-manager: docs/contributor/pocs/otlp-logs.md

## Log Record Parsing

### Scope and Goals

When integrating an OTLP-compliant logging backend, applications can either ingest their logs directly or emit them to STDOUT and use a log collector to process and forward the logs.
The first part of this PoC evaluates how to configure the OpenTelemetry Collector's [filelog receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver) to transform structured JSON logs emitted by Kubernetes workloads to STDOUT, and subsequently to the [OTLP logs data model](https://opentelemetry.io/docs/specs/otel/logs/data-model/).
OpenTelemtry Collector should move JSON attributes to the **attributes** map of the log record, extract other fields like **severity** or **timestamp**, write the actual log message to the **body** field, and add any missing information to ensure that the **attributes** and **resource** attributes comply with the semantic conventions.

This PoC does not cover logs ingested by the application using the OTLP protocol. We assume that the application already fills the log record fields with the intended values.

## Setup for Log Record Parsing

> We created a Helm values file for the `open-telemetry/opentelemetry-collector` chart that parses and transforms container logs in the described way. We use an SAP Cloud Logging instance as the OTLP compliant logging backend. To deploy the setup, follow these steps:

Reviewer: ```suggestion
We created a Helm values file for the `open-telemetry/opentelemetry-collector` chart that parses and transforms container logs in the described way. We use an SAP Cloud Logging instance as the OTLP-compliant logging backend. To deploy the setup, follow these steps:
```

> 

Reviewer: change line 13 to `## Setup for Log Record Parsing`

Author: Done

The OpenTelemetry Collector setup is able to extract the log message from different attributes, depending on their presence. This means that it is possible to support different logging libraries.

Non-JSON logs are preserved in the **body** field until the enrichment with resource attributes is completed.

## 2. Buffering and Backpressure

### Scope and Goals

After evaluating the filelog receiver configuration in the first part of the PoC, we want to test the buffering and backpressure capabilities of the OpenTelemetry Collector. The OpenTelemetry based logging solution should give similar resilience and guarantees about log delivery as the current logging solution.

## Setup

> We split the OpenTelemetry Collector for log processing to an agent (DaemonSet) and a gateway (StatefulSet). The agent uses the same configuration as shown in the first part of the PoC to read logs from the host file-system and converts them to the OTLP format, while the gateway adds Kubernetes metadata and ensures that no logs are lost in the case of a backend failure.

Reviewer: ```suggestion
We split the OpenTelemetry Collector for log processing to an agent (DaemonSet) and a gateway (StatefulSet). The agent uses the same configuration as shown in [Log Record Parsing](#log-record-parsing) to read logs from the host file system, and converts them to the OTLP format, while the gateway adds Kubernetes metadata and ensures that no logs are lost if the backend fails.
```

> The figure below shows the different plugins that are configured in the processing pipeline. Important to mention is the use of the batch processor in the gateway, which introduces asynchronicity to the pipeline and causes that backpressure is not propagated back to the agent. To minimize the risk of log loss due to the batch processors properties, a persistent exporter queue was introduced in the gateway, which uses a persistent volume to buffer logs in case of a backend failure.

Reviewer: ```suggestion
The following figure shows the different plugins that are configured in the processing pipeline. Note the use of the batch processor in the gateway, which introduces asynchronicity to the pipeline and causes that backpressure is not propagated back to the agent. To minimize the risk of log loss due to the batch processors properties, a persistent exporter queue was introduced in the gateway, which uses a persistent volume to buffer logs in case of a backend failure.
```
Excellent use of relative clauses with the correct relative pronoun and punctuation.

![Otel Collector Setup](./assets/otlp-logs.drawio.svg)

To deploy the OpenTelemetry Collector agent and gateway, perform the following steps:

> 1. Create a SAP Cloud Logging instance as described above

Reviewer: ```suggestion
1. Create a SAP Cloud Logging instance as described in [Setup for Log Record Parsing](#setup-for-log-record-parsing).
```

1. Create a persistent volume claim (PVC):

   ```bash
   kubectl apply -n otel-logging -f ./assets/otel-gateway-pvc.yaml
   ```

1. Deploy the gateway:

   ```bash
   helm repo add open-telemetry https://open-telemetry.github.io/opentelemetry-helm-charts
   helm install -n otel-logging logging open-telemetry/opentelemetry-collector -f ./assets/otel-log-gateway-values.yaml
   ```

1. Deploy the agent:

   ```bash
   helm install -n otel-logging logging open-telemetry/opentelemetry-collector -f ./assets/otel-log-agent-values.yaml
   ```

### Evaluation

> To evaluate the buffering and backpressure capabilities of the described OpenTelemetry Collector setup, we tested the following scenarios and observed the described behavior:

Reviewer: ```suggestion
To evaluate the buffering and backpressure capabilities of the described OpenTelemetry Collector setup, we tested the following scenarios and observed the respective behavior:
```

* **Outage of the OTLP backend**

>   Log records cannot be shipped from the gateway to the OTLP backend (SAP Cloud Logging). Once the configured queue limit is reached, log records will be dropped from the queue. The enqueue errors are not propagated back to other pipeline elements due to the asynchronicity introduced by the batch processor.

Reviewer: ```suggestion
  Log records cannot be shipped from the gateway to the OTLP backend (SAP Cloud Logging). When the configured queue limit is reached, log records are dropped from the queue. The enqueue errors are not propagated back to other pipeline elements because of the asynchronicity introduced by the batch processor.
```

* **Broken connectivity between the agent and the gateway**

>   Log records cannot be exported by the agent to the gateway using the OTLP protocol. The exporter queue on the agent will up to its maximum size and then start rejecting new records. This enqueue error is propagated to the filelog receiver, which eventually stops reading new logs. Log loss is avoided until the log retention of the kubelet removes old logs.

Reviewer 2: ```suggestion
 The exporter queue on the agent will buffer up to its maximum size and then start rejecting new records.
```
The sentence looks incomplete.

>   Log records cannot be exported by the agent to the gateway using the OTLP protocol. The exporter queue on the agent will up to its maximum size and then start rejecting new records. This enqueue error is propagated to the filelog receiver, which eventually stops reading new logs. Log loss is avoided until the log retention of the kubelet removes old logs.

Reviewer 2: ```suggestion
  Log records cannot be exported by the agent to the gateway using the OTLP protocol. The exporter queue on the agent will buffer up to its maximum size and then start rejecting new records. This enqueue error is propagated to the filelog receiver, which eventually stops reading new logs. Log loss is avoided until the log retention of the kubelet removes old logs.
```

>   Log records cannot be exported by the agent to the gateway using the OTLP protocol. The exporter queue on the agent will buffer up to its maximum size and then start rejecting new records. This enqueue error is propagated to the filelog receiver, which eventually stops reading new logs. Log loss is avoided until the log retention of the kubelet removes old logs.

Reviewer: ```suggestion
  Log records cannot be exported by the agent to the gateway using the OTLP protocol. The exporter queue on the agent buffers up to its maximum size and then starts rejecting new records. This enqueue error is propagated to the filelog receiver, which eventually stops reading new logs. Log loss is avoided until the log retention of the kubelet removes old logs.
```

### Conclusions

> The evaluation of the two failure scenarios showed that the OpenTelemetry Collector can similar guarantees about the prevention of log loss as the current Fluent Bit setup. When using a batch processer, using a persistent output queue and with that increasing the queue capacity, helps to prevent data loss. Splitting that processing pipeline to agent and gateway allows to use a PVC for the exporter queue and with that give it a large capacity without the risk that the node file-system fills up.

Reviewer 2: ```suggestion
The evaluation of the two failure scenarios showed that the OpenTelemetry Collector can have similar guarantees about the prevention of log loss as the current Fluent Bit setup. When using a batch processor, using a persistent output queue and with that increasing the queue capacity, helps to prevent data loss. Splitting that processing pipeline to agent and gateway allows to use a PVC for the exporter queue and with that give it a large capacity without the risk that the node file-system fills up.
```

May be re-writing this sentence would help:  `When using a batch processor, using a persistent output queue and with that increasing the queue capacity, helps to prevent data loss.` its difficult to follow

> The evaluation of the two failure scenarios showed that the OpenTelemetry Collector can have similar guarantees about the prevention of log loss as the current Fluent Bit setup. When using a batch processor, using a persistent output queue and with that increasing the queue capacity, helps to prevent data loss. Splitting that processing pipeline to agent and gateway allows to use a PVC for the exporter queue and with that give it a large capacity without the risk that the node file-system fills up.

Reviewer: ```suggestion
The evaluation of the two failure scenarios shows that the OpenTelemetry Collector can have similar guarantees about the prevention of log loss as the current Fluent Bit setup. When using a batch processor, we can prevent data loss with a persistent output queue, which increases the queue capacity. By splitting that processing pipeline to agent and gateway, we can use a PVC for the exporter queue, which provides large capacity without the risk that the node file system fills up.
```

During the evaluation, the following potential problems and risks have been identified:

* The persistent queue of the OpenTelemetry Collector is still in alpha state and might not be suitable yet for production use.

> * The queue capacity is configured by the number of batches. A storage capacity based limitation is not possible. This makes it hard to give exact guarantees about the stored logs before data loss.

Reviewer: ```suggestion
* The queue capacity is configured by the number of batches. A limitation based on storage capacity is impossible. This makes it hard to give exact guarantees about the stored logs before data loss.
```

> * Once allocated, the utilized storage space of the persistent queue never shrinks again. This is not a problem as long as a dedicated PVC is used for the queue, but makes it less suitable to be stored on the node's host file system.

Reviewer: ```suggestion
* After it is allocated, the utilized storage space of the persistent queue never shrinks again. This is not a problem as long as a dedicated PVC is used for the queue, but makes it less suitable to be stored on the node's host file system.
```

## Setup for Buffering and Backpressure

We split the OpenTelemetry Collector for log processing to an agent (DaemonSet) and a gateway (StatefulSet). The agent uses the same configuration as shown in [Log Record Parsing](#log-record-parsing) to read logs from the host file system, and converts them to the OTLP format, while the gateway adds Kubernetes metadata and ensures that no logs are lost if the backend fails.

The following figure shows the different plugins that are configured in the processing pipeline. Note the use of the batch processor in the gateway, which introduces asynchronicity to the pipeline and causes that backpressure is not propagated back to the agent. To minimize the risk of log loss due to the batch processors properties, a persistent exporter queue was introduced in the gateway, which uses a persistent volume to buffer logs in case of a backend failure.

> ![Otel Collector Setup](./assets/otlp-logs.drawio.svg)

Reviewer: This diagram looks weird when rendered: 
<img width="1036" alt="image" src="https://github.com/kyma-project/telemetry-manager/assets/76950046/8900a033-6710-49bd-a5a8-e388437bdc4c">

Consider changing the colors and background; for example, according to our [guidelines](https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/09-diagrams.md).
